### ISSUE TITLE
Public constructors for wrapper types

### ISSUE THREAD

**Issue by samarsha** [TheDan64/inkwell#366]
**Is your feature request related to a problem? Please describe.**
It's difficult to interoperate between Inkwell and llvm-sys. This is a problem because Inkwell is not complete, so some things are difficult or impossible to do without calling LLVM APIs directly.

The get_ref() method from LLVMReference gets me only halfway there. If I use an LLVM pointer to create a new LLVM object, like an LLVMValueRef, it's not possible for me to put it back in an Inkwell wrapper. So I can't fill in gaps in Inkwell's functionality by unwrapping, using llvm-sys, and rewrapping. Everything I do with the value afterwards has to be through llvm-sys, even if I would be able to use Inkwell otherwise.

**Describe the solution you'd like**
All `pub(crate)` constructors like

```rust
impl<'ctx> AnyValueEnum<'ctx> {
    pub(crate) unsafe fn new(value: LLVMValueRef) -> Self { ... }
}
```

should be made `pub`.

**Describe possible drawbacks to your solution**
It weakens encapsulation, but I think it's worth it for pragmatism, and Inkwell has already started down that path with get_ref().

**Describe alternatives you've considered**
The only alternative I see is forking Inkwell or using 100% llvm-sys instead of a mixture of Inkwell and llvm-sys.

**Comment by TheDan64** [comment]
It seems like the better solution is to add needed functionality from llvm sys into inkwell. Contributions are always welcome

**Comment by TheDan64** [comment]
For context, inkwell was never intended to interoperate with llvm-sys, but to purely be a wrapper ontop of it

**Comment by samarsha** [comment]
> It seems like the better solution is to add needed functionality from llvm sys into inkwell. Contributions are always welcome

It's definitely good if Inkwell becomes better so that I need llvm-sys less, but I don't agree that it's better for Inkwell to prevent me from using llvm-sys (or any other LLVM library) when needed. By keeping the constructors pub(crate), Inkwell is artificially making itself an all-or-nothing proposition: either I use Inkwell for everything, and wait for PRs for functionality that isn't supported yet, or I don't use Inkwell at all. LLVM changes rapidly and I think it's realistic to admit that Inkwell can't always keep up.

I also think that, even an ideal world where Inkwell always supported 100% of LLVM's functionality perfectly, allowing Inkwell to work with LLVM objects that didn't originate from Inkwell is still good. You might want to combine Inkwell with another LLVM library that works with LLVM objects directly, and doesn't know about Inkwell types.

I have a situation like that with an LLVM library I'm making that has two components, a Rust crate that depends on Inkwell, and a Python package. The Python package is a native module that depends on the Rust crate and uses PyO3 to expose LLVM to Python. In Python, I don't really benefit from Inkwell's lifetime safety since it all goes in the Python heap anyway. It would actually be more convenient if I could use LLVM pointers in my Python classes and then convert back to Inkwell when needed, but I can't.

> For context, inkwell was never intended to interoperate with llvm-sys, but to purely be a wrapper ontop of it

I'm not sure I understand the difference. I agree it's a wrapper, but my concern is that it doesn't let me wrap LLVM objects that came from somewhere else, only unwrap Inkwell types into LLVM objects.

**Comment by TheDan64** [comment]
Fair enough

**Comment by samarsha** [comment]
This issue isn't fixed by #372. That's a step in the right direction, but there's lots of other Inkwell types that don't expose their constructors: Context, Module, Builder, everything in inkwell::values, etc.

**Comment by gavrilikhin-d** [comment]
Why would you want to create inkwell builder from foreign code?
